base/promotion.jl

```diff
@@ -2,6 +2,10 @@
 
 ## type join (closest common ancestor, or least upper bound) ##
 
+isspecial(::Type) = false
+isspecial(::Type{Void}) = true
+isspecialpair(a, b) = isconcrete(a) && isconcrete(b) && (isspecial(a) || isspecial(b))
```

Comment:
If you're intending for people to override this, `typejoin` won't be pure anymore (currently marked as such)

Reply:
Even if people are only supposed to override this with `isspecial(::Type{Null}) = true`? `@pure` is still a mystery to me (is the problem due to the addition of methods after the fact?).

Reply:
> is the problem due to the addition of methods after the fact?

Yes.